designs/2024-multithread-linting/README.md

```diff
@@ -0,0 +1,780 @@
+
+            if (unserializableKeys.length) {
+                throw new Error(`Unserializable options not allowed with "concurrency"\nFound unserializable options: ${
+                    new Intl.ListFormat('en-US').format(unserializableKeys.map(key => `"${key}"`))}
+                `);
+            }
+        }
+```
+
+The `isSerializable` function is defined from the requirements for serializable objects.
+Trivially, one could call `structuredClone()` or [`v8.serialize()`](https://nodejs.org/docs/latest-v18.x/api/v8.html#v8serializevalue) on a specified value and if an error is caught, assume that the value is unserializable.
+If there is a more correct implementation we could use that one instead.
+`JSON.stringify` cannot be used because it simply ignores unserializable values.
+
+### `ESLint#lintFiles()`
+
+The core of this proposal aims at providing an alternative implementation for the `ESLint#lintFiles()` method, or more exactly for the inner part of that method, after files are enumerated and before the cache is saved.
+See the [_relevant code_](https://github.com/eslint/eslint/blob/v9.15.0/lib/eslint/eslint.js#L747-L839) in `lib/eslint/eslint.js`.
+
+The new multithread implementation will exist alongside the current, single-thread implementation, and it will only be used when multithreading mode is enabled.
+
+In the multithread implementation `lintFiles()` launches as many worker threads as specified by the `concurrency` option, but no more than the number of enumerated files.
+All worker threads receive the same parameters from the controlling thread:
+* the list of enumerated file paths
+* a reference to shared memory containing the index of the next file to be processed in the list (a `TypedArray` with an underlying `SharedArrayBuffer`)
+* only one of the following:
+  * the serializable `ESLint` options
+  * the URL of an [option module](#option-modules)
+  * _possibly_ the parsed CLI options, depending on the chosen [solution for CLI options](#solution-for-cli-options)
+
+Each worker thread repeatedly reads and lints a single file until all files have been linted or an abort signal is triggered.
+Errors in worker threads are not caught: they will be intercepted as error events by the controlling thread where they will trigger an abort signal that causes all other threads to exit.
+
+The controlling thread itself does not lint any files: it waits until all files are linter or an error occurs.
+When a worker thread terminates successfully it submits a list of `LintReport`s to the controlling thread. Each result is enriched with the index of the associated file.
+The controlling thread collects the lists of `LintReport`s and merges them in the expected order, then populates `lintResultCache` with cache information (**note**: this is different from the single-thread implementation where the cache is populated progressively as files are being linted).
```

Comment:
> merges them in the expected order

I presume this is part of the reason for waiting and batching results rather than streaming them. Is "expected order" something that ESLint guarantees? Or requires preserving in a meaningful way? I could imagine if the user passes in a concrete list of files that possibly they'd want to see the results in exactly that order, but personally I also think this does not matter, and I don't believe any of the existing parallelism implementations attempt to preserve this behavior either (i.e., a result contains the file name and location, so it doesn't matter if its index matches the original file list order).

Reply:
> Is "expected order" something that ESLint guarantees? Or requires preserving in a meaningful way? I could imagine if the user passes in a concrete list of files that possibly they'd want to see the results in exactly that order

The order of lint results depends on how files are enumerated before linting starts. This does not necessarily reflect the order specified by a user, but it's stable in the sense that it doesn't change across executions. For multithread mode, the order of lint results will be the same as for single-thread mode.